syntax = "proto3";

import "google/protobuf/timestamp.proto";
import "ledger.proto";

package pieledger;

message TransactionQueryRequest {
  int32 page_number = 1;
  int32 result_per_page = 2;
  repeated string guids = 3;
  Account account = 4;
}

message SplitQueryRequest {
  int32 page_number = 1;
  int32 result_per_page = 2;
  Account account = 5;
  google.protobuf.Timestamp from_datetime = 3;
  google.protobuf.Timestamp to_datetime = 4;
  oneof by_action {
    SplitAction standard_action = 6;
    string custom_action = 7;
  }
}

message TransactionAlterRequest {
  Transaction transaction = 1;
  repeated Split append_splits = 2;
}

service PieLedger {
  rpc FindOrCreateAccount(Account) returns (Account);
  rpc AlterAccount(Account) returns (Account);

  rpc CreateTransaction(Transaction) returns (Transaction);
  rpc FindTransactions(TransactionQueryRequest) returns (stream Transaction);
  rpc AlterTransaction(TransactionAlterRequest) returns (Transaction);

  rpc FindSplits(SplitQueryRequest) returns (stream Split);
}
